File: internal/app/repo/postgres/admin_repo.go
package postgres

import (
	"context"
	"database/sql"
	"fmt"
	"log/slog"
	"mahir-trade-be/internal/app/models"
	"mahir-trade-be/internal/app/repo/postgres/queries"
	"strings"

	"go.uber.org/dig"
)

type (
	AdminRepo interface {
		FindByUsername(ctx context.Context, username string) (admin models.Admin, err error)
		CreateAdmin(ctx context.Context, req models.Admin) (id int, err error)
		SoftDeleteAdmin(ctx context.Context, id int64, operator string) (err error)
	}

	AdminRepoImpl struct {
		dig.In

		*sql.DB
	}
)

func NewAdminRepo(impl AdminRepoImpl) AdminRepo {
	return &impl
}

func (a *AdminRepoImpl) FindByUsername(ctx context.Context, username string) (admin models.Admin, err error) {
	err = a.QueryRowContext(ctx, queries.QueryFindByEmail, username).Scan(&admin.AdminID, &admin.UUID, &admin.Email, &admin.Username, &admin.Password)
	return admin, err
}

func (a *AdminRepoImpl) CreateAdmin(ctx context.Context, req models.Admin) (id int, err error) {
	var rows *sql.Rows

	rows, err = a.QueryContext(ctx, queries.QueryCreateAdmin, req.Email, req.Username, req.Password)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while CreateAdmin err: %v", err.Error()))
		if strings.Contains(err.Error(), "duplicate key value violates unique constraint") {
			return id, fmt.Errorf("email or username already exist")
		}

		return id, err
	}

	defer rows.Close()

	for rows.Next() {
		err = rows.Scan(&id)
		if err != nil {
			return id, err
		}
	}

	return id, err
}

func (a *AdminRepoImpl) SoftDeleteAdmin(ctx context.Context, id int64, operator string) (err error) {
	_, err = a.ExecContext(ctx, queries.QuerySoftDeleteAdmin, operator, operator, id)
	return err
}

File: src/repo/admin_repo.ts
import { Pool } from "pg";
import { Admin } from "../models/admin.model";

import db from "../db/db.config";
import { AdminQueries } from "./queries/admin_queries";

export class AdminRepository {
  private pg: Pool;

  constructor() {
    this.pg = db;
  }

  // --- FIND BY USERNAME ---
  async findByUsername(username: string): Promise<Admin | null> {
    try {
      const result = await this.pg.query(AdminQueries.FindByEmail, [username]);

      if (result.rowCount === 0) {
        return null; // sama seperti return kosong di Go
      }

      return result.rows[0] as Admin;
    } catch (err: any) {
      console.error(`[repo][admin][findByUsername] error: ${err.message}`);
      throw err;
    }
  }

  // --- CREATE ADMIN ---
  async createAdmin(req: Admin): Promise<number> {
    try {
      const result = await this.pg.query(AdminQueries.CreateAdmin, [
        req.email,
        req.username,
        req.password,
      ]);

      if (result.rowCount === 0) {
        throw new Error("failed to create admin");
      }

      return result.rows[0].id as number; // sama dengan rows.Scan(&id) di Go
    } catch (err: any) {
      if (
        err.message.includes("duplicate key value") &&
        err.message.includes("unique constraint")
      ) {
        throw new Error("email or username already exist");
      }
      console.error(`[repo][admin][createAdmin] error: ${err.message}`);
      throw err;
    }
  }

  // --- SOFT DELETE ADMIN ---
  async softDeleteAdmin(id: number, operator: string): Promise<boolean> {
    try {
      const result = await this.pg.query(AdminQueries.SoftDeleteAdmin, [
        operator,
        operator,
        id,
      ]);

      return (result.rowCount ?? 0) > 0;
    } catch (err: any) {
      console.error(`[repo][admin][softDeleteAdmin] error: ${err.message}`);
      throw err;
    }
  }
}
